### ISSUE TITLE
Default value for json/jsonb column leads to crash

### ISSUE THREAD

**Issue by Dador** [bizley/yii2-migration#50]
Hello.

Yii converts default value for json/jsonb to array and seems like yii2-migration doesn't expects this.

Log:
```
PHP Notice 'yii\base\ErrorException' with message 'Array to string conversion'

in vendor\bizley\migration\src\table\TableColumn.php:149
```

Quick & dirty way to fix it is to add `elseif` [to TableColumn.php:149](https://github.com/bizley/yii2-migration/blob/3.2.0/src/table/TableColumn.php#L149):
```php
elseif (is_array($this->default)) {
    $this->definition[] = "defaultValue('" . $this->escapeQuotes(json_encode($this->default)) . "')";
} 
```
Though probably `array` used in yii2 not only for json/jsonb, but also for other column types, so maybe there is better way to fix it.

P.S. Also there is issue with default value `{}` (object), yii2 convert it to php empty array `[]` and json_encode convert it to `[]`, so it becomes an array instead of object. There is no difference when you use json in PHP only, but for other languages (and in database) that could be a problem. Though that's already beyond control of this extension.

**Comment by bizley** [comment]
To be honest I've never thought about giving default json value as actual array because for me it was more of Yii implementation of this column output. But I don't see a reason of not adding this. Thank you.

**Comment by bizley** [comment]
I've got problem with this one.

1) Is Yii allowing providing arrays in `defaultValue()` at all? Or at least for DBMS supporting it like PostgreSQL?

2) If first question answer is no then is providing Json-encoded string enough? Or should it be switched to `defaultExpression()`? With `JsonExpression` or not?

Could anyone help?

**Comment by Dador** [comment]
> Is Yii allowing providing arrays in defaultValue() at all? Or at least for DBMS supporting it like PostgreSQL?

I guess Yii doesn't support providing array value (because [here](https://github.com/yiisoft/yii2/blob/master/framework/db/ColumnSchemaBuilder.php#L337) object type casted to string and array processed like a string).

> If first question answer is no then is providing Json-encoded string enough? Or should it be switched to defaultExpression()? With JsonExpression or not?

Yes, providing json-encoded string is enough. I think there is no need to use defaultExpression.

For example:
```
'data' => $this->json()->notNull()->defaultValue('[]'),
```
or
```
'data' => $this->json()->notNull()->defaultValue('{"name": "value"}'),
```

So fix from the first message does works. I just thought that maybe there is better place to put it (like TableColumnJson instead of TableColumn).

> Should the extension do anything different in case of generalSchema=1 option or not?

Probably it should skip specifying default value for MySQL < 8.0.13 (if that's possible) (though that's a bit non-consistent behavior).
Or it could specify default value only for PostgreSQL and only in case of empty array (seems like in MySQL empty array is standart default value for JSON) (that's quite limited but consistent).

PostgreSQL supports JSON and default value for JSON since version 9.2 (released 7 years ago in 2012).

MySQL supports JSON since version 5.7.8 and only since 8.0.13 (released 6 months ago) it supports default value for JSON. Also looks like `[]` is default value for json in MySQL anyway, so there is no need (in case of MySQL) to explicitly specify default value `[]`.

**Comment by bizley** [comment]
I decided to keep json-encoded array as default for `defaultValue()` in case the source column is array.

Anyway, could you check the current dev-master? I think I've managed to fix the problem.

The bad news is that by digging into PostgreSQL way of altering tables I learned that I don't have special cases of altering covered in this extension. Could you share your thoughts about #52 ?

**Comment by bizley** [comment]
Fixed in [3.3.0](https://github.com/bizley/yii2-migration/releases/tag/3.3.0) / [2.6.0](https://github.com/bizley/yii2-migration/releases/tag/2.6.0)
